### PR TITLE
Reduce reliance on promotion operations via container typing

### PR DESCRIPTION
**Pull request by timholy**
While building Julia, we ~have~had a *lot* of calls to promote, often with abstract types. This short-circuits most of these by declaring the container type at the outset.

This has one user-visible outcome, improved inference for `walkdir`.

After my complete PR series is submitted, I'll post some benchmarks here on the impact on building Julia itself. Obviously no one PR accounts for the whole thing, but this one may be as good a place as any to deposit this info.

### CODE REVIEW

**Review by quinnj — APPROVED** [review]
Very cool stuff!

**Comment by timholy** [comment]
master:
```julia
Sysimage built. Summary:
Total ───────  44.593185 seconds 
Base: ───────  18.240267 seconds 40.9037%
Stdlibs: ────  26.351632 seconds 59.0934%
    JULIA usr/lib/julia/sys-o.a
Generating precompile statements... 30/30
Executing precompile statements... 1771/1782
Precompilation complete. Summary:
Total ───────  85.293917 seconds
Generation ──  58.616914 seconds 68.7234%
Execution ───  26.677004 seconds 31.2766%
    LINK usr/lib/julia/sys.so
tim@diva:~/src/julia-master$ ls -l usr/lib/julia/sys.so 
-rwxr-xr-x 1 tim holy 148227920 Aug 17 15:28 usr/lib/julia/sys.so
```

On a branch that merges all the PRs I've submitted today:
```julia
Sysimage built. Summary:
Total ───────  42.541928 seconds 
Base: ───────  17.358497 seconds 40.8033%
Stdlibs: ────  25.181829 seconds 59.193%
    JULIA usr/lib/julia/sys-o.a
Generating precompile statements... 30/30
Executing precompile statements... 1652/1663
Precompilation complete. Summary:
Total ───────  79.426395 seconds
Generation ──  55.432791 seconds 69.7914%
Execution ───  23.993604 seconds 30.2086%
    LINK usr/lib/julia/sys.so
tim@diva:~/src/julia-master$ ls -l usr/lib/julia/sys.so
-rwxr-xr-x 1 tim holy 144046664 Aug 17 15:44 usr/lib/julia/sys.so
```

So:
- about 2s faster to build up through precompilation
- about 6s faster to precompile, with 100 fewer MethodInstances
- a sysimg that is 3% smaller

Not too shabby for a couple days of digging with Cthulhu.

File: base/version.jl [hunk 1]
```diff
@@ -110,15 +112,15 @@ function VersionNumber(v::AbstractString)
     m = match(VERSION_REGEX, v)
     m === nothing && throw(ArgumentError("invalid version string: $v"))
     major, minor, patch, minus, prerl, plus, build = m.captures
-    major = parse(VInt, major)
+    major = parse(VInt, major::AbstractString)
```
Comment (KristofferC) [inline base/version.jl:115]: I think this can be `SubString{String}` since `match` is only usable with `String` and `SubString{String}`.

**Review by KristofferC — APPROVED** [review]

File: base/version.jl (hunk 1 above)
Comment (timholy) [reply base/version.jl:115]: Oops, that was leftover from an experiment. The important change is the `VerTuple` assert.

**Pull request merged**